swift/extractor/translators/ExprTranslator.cpp

```diff
@@ -488,6 +488,26 @@ TrapLabel<ArgumentTag> ExprTranslator::emitArgument(const swift::Argument& arg)
   return entry.id;
 }
 
+TrapLabel<KeyPathComponentTag> ExprTranslator::emitKeyPathComponent(
+    const swift::KeyPathExpr::Component& component) {
+  auto entry = dispatcher.createUncachedEntry(component);
+  entry.kind = static_cast<int>(component.getKind());
+  if (auto subscript_args = component.getSubscriptArgs()) {
+    for (const auto& arg : *subscript_args) {
+      entry.subscript_arguments.push_back(emitArgument(arg));
+    }
+  }
+  if (component.getKind() == swift::KeyPathExpr::Component::Kind::TupleElement) {
+    entry.tuple_index = static_cast<int>(component.getTupleIndex());
+  }
+  if (component.hasDeclRef()) {
+    entry.decl_ref = dispatcher.fetchLabel(component.getDeclRef().getDecl());
+  }
```

Comment:
No, I've made an attempt to handle everything already. Is there something I've missed?

In terms of the enumeration specifically, we store the kind as an integer (`entry.kind = static_cast<int>(component.getKind());`); schema.py then documents that some of the kinds won't occur in a DB (e.g. `CodeCompletion` is only for IDEs and `DictionaryKey` is for ObjC-only `#keyPath`). The kinds we do support, I believe we extract everything for.

Note that `Component` isn't a union, so we extract a lot of things without checking whether they apply to a given kind. We make an exception for tuple index because the accessor for that asserts it's a tuple, whereas other accessors have fallbacks.

Reply:
Oh, I thought Swift also generated `DictionaryKey` components. But if that's ObjC-only then I'm perfectly happy with that one.

How about `OptionalForce`, `OptionalChain`, and `OptionalWrap`? For dataflow purposes we (since https://github.com/github/codeql/pull/12133) handle `T` <-> `T?` conversions, but that relies on the conversions being in the DB. Do we not need explicit handling of those components here? Or do we get that automatically from our existing visitors?

Reply:
[This](https://github.com/apple/swift/blob/d563b587ddc59ef53b19022121b537a8edfa62f8/include/swift/AST/Expr.h#L5655) makes `DictionaryKey` sound `#keyPath`-only, and from what I can tell `#keyPath` is ObjC only.

`OptionalForce` and `OptionalChain` don't need any special handling in the extractor - we can tell which one it is from the `kind` (that we extract for all components) and `component_type` also tells us the post-conversion type. `OptionalWrap` is a bit awkward in that it doesn't have a source location (since it's always implicitly inserted by the compiler), but it is present and ordered correctly in the `KeyPathExpr::components` list; it's also identifiable by its kind and also has a `component_type`.

Reply:
Ah, right. Yeah, I guess we can get those conversions from the component type. Awesome 👍